### PR TITLE
Modified refresh client to calculate the shortest token expiry

### LINKED ISSUE #1189: `Refresh` client should calculate the shortest token expiry time
## Expected Behavior
`Refresh` client should calculate the shortest token expiry in the chain. This is required for refresh feature to work correctly in the case when there're multiple mesh elements with different token expiry (e.g. in inter-domain case).

## Current Behavior
The shortest token expiry is not calculated.

## Context
Bug is opened based on discussion in #1185.

## Proposed Solution
`Refresh` client can iterate over path segments and calculate the shortest token expiry.

### PR DESCRIPTION
**Pull request by Sol-0**
Signed-off-by: Oleg Solodkov <oleg.solodkov@xored.com>

## Description
Refresh client calculates the shortest token expiry and sets refresh time according to that value

## Issue link
Fixes #1189

## How Has This Been Tested?
- [x] Added unit testing to cover
- [ ] Tested manually
- [ ] Tested by integration testing
- [ ] Have not tested

## Types of changes
- [x] Bug fix
- [ ] New functionallity
- [ ] Documentation
- [ ] Refactoring
- [ ] CI

### CODE REVIEW

File: pkg/networkservice/common/refresh/client.go [hunk 1]
```diff
@@ -108,16 +109,23 @@ func (t *refreshClient) Close(ctx context.Context, conn *networkservice.Connecti
 func after(ctx context.Context, conn *networkservice.Connection) (time.Duration, error) {
 	clockTime := clock.FromContext(ctx)
 
-	expireTime, err := ptypes.Timestamp(conn.GetCurrentPathSegment().GetExpires())
-	if err != nil {
-		return 0, errors.WithStack(err)
-	}
+	var minTimeout = time.Duration(math.MaxInt64)
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/refresh/client.go:112]:
```suggestion
	var minTimeout time.Duration
```

File: pkg/networkservice/common/refresh/client.go [hunk 2]
```diff
@@ -108,16 +109,23 @@ func (t *refreshClient) Close(ctx context.Context, conn *networkservice.Connecti
...
+	for _, seg := range conn.GetPath().GetPathSegments() {
+		expireTime, err := ptypes.Timestamp(seg.GetExpires())
+		if err != nil {
+			return 0, errors.WithStack(err)
+		}
 
-	timeout := clockTime.Until(expireTime)
-	log.FromContext(ctx).Infof("expiration after %s at %s", timeout.String(), expireTime.UTC())
+		timeout := clockTime.Until(expireTime)
+		log.FromContext(ctx).Infof("expiration after %s at %s", timeout.String(), expireTime.UTC())
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/refresh/client.go:120]:
```suggestion
```

File: pkg/networkservice/common/refresh/client.go [hunk 3]
```diff
@@ -108,16 +109,23 @@ func (t *refreshClient) Close(ctx context.Context, conn *networkservice.Connecti
...
-	if timeout <= 0 {
-		return 1, nil
+		if timeout <= 0 {
+			return 1, nil
+		}
+
+		if timeout < minTimeout {
+			minTimeout = timeout
+		}
 	}
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/refresh/client.go:129]:
```suggestion
	}
	log.FromContext(ctx).Infof("expiration after %s at %s", timeout.String(), expireTime.UTC())
        if timeout <= 0 {
	        return 1, nil
        }
```

File: pkg/networkservice/common/refresh/client.go [hunk 4]
```diff
@@ -108,16 +109,23 @@ func (t *refreshClient) Close(ctx context.Context, conn *networkservice.Connecti
...
-	timeout := clockTime.Until(expireTime)
-	log.FromContext(ctx).Infof("expiration after %s at %s", timeout.String(), expireTime.UTC())
+		timeout := clockTime.Until(expireTime)
+		log.FromContext(ctx).Infof("expiration after %s at %s", timeout.String(), expireTime.UTC())
 
-	if timeout <= 0 {
-		return 1, nil
+		if timeout <= 0 {
+			return 1, nil
+		}
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/refresh/client.go:124]:
```suggestion
```

**Review by denis-tingaikin — CHANGES_REQUESTED** [review]

File: pkg/networkservice/common/refresh/client.go [hunk 5]
```diff
@@ -108,16 +109,23 @@ func (t *refreshClient) Close(ctx context.Context, conn *networkservice.Connecti
 func after(ctx context.Context, conn *networkservice.Connection) (time.Duration, error) {
 	clockTime := clock.FromContext(ctx)
 
-	expireTime, err := ptypes.Timestamp(conn.GetCurrentPathSegment().GetExpires())
-	if err != nil {
-		return 0, errors.WithStack(err)
-	}
+	var minTimeout = time.Duration(math.MaxInt64)
+	for _, seg := range conn.GetPath().GetPathSegments() {
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/refresh/client.go:113]:
```suggestion
	for _, segment := range conn.GetPath().GetPathSegments() {
```

File: pkg/networkservice/common/refresh/client_test.go [hunk 6]
```diff
@@ -302,6 +325,82 @@ func TestRefreshClient_NoRefreshOnFailure(t *testing.T) {
...
+
+	testData := []struct {
+		Chain                     []time.Duration
+		ExpectedRefreshTimeoutMax time.Duration
+		ExpectedRefreshTimeoutMin time.Duration
+	}{
+		{
+			Chain:                     []time.Duration{time.Hour},
+			ExpectedRefreshTimeoutMax: 20*time.Minute + time.Second,
+			ExpectedRefreshTimeoutMin: 20*time.Minute - time.Second,
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/refresh/client_test.go:342]: Please use const time range.
```suggestion
			ExpectedRefreshTimeout: 20*time.Minute
```

Comment (denis-tingaikin) [reply pkg/networkservice/common/refresh/client_test.go:342]: Apply this for other places.

File: pkg/networkservice/common/refresh/client_test.go [hunk 7]
```diff
@@ -302,6 +325,82 @@ func TestRefreshClient_NoRefreshOnFailure(t *testing.T) {
...
+		{
+			Chain:                     []time.Duration{200 * time.Millisecond, 1 * time.Minute, 100 * time.Millisecond, time.Hour},
+			ExpectedRefreshTimeoutMax: 100*time.Millisecond/3. + 30*time.Millisecond,
+			ExpectedRefreshTimeoutMin: 100*time.Millisecond/3. - 30*time.Millisecond,
+		},
+	}
+
+	timeNow, err := time.Parse("2006-01-02 15:04:05", "2009-11-10 23:00:00")
+	require.NoError(t, err)
+
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/refresh/client_test.go:363]: `time.Date()` is not returning error.

```suggestion
	timeNow := time.Date(/*todo*/)

```

**Comment by denis-tingaikin** [comment]
@edwarnicke We are thinking to solve https://github.com/networkservicemesh/sdk/issues/1189 with this PR.
Please have a look and share your thoughts.

**Review by Mixaster995 — APPROVED** [review]

File: pkg/networkservice/common/refresh/client.go (hunk 1 above)
Comment (Sol-0) [reply pkg/networkservice/common/refresh/client.go:112]: If we don't initialize `minTimeout` with max possible duration, it will always be `0`.

Comment (denis-tingaikin) [reply pkg/networkservice/common/refresh/client.go:112]: Please fix the case when there no segments.

**Review by denis-tingaikin — APPROVED** [review]
LGTM

@edwarnicke Could you have a look?

File: pkg/networkservice/common/refresh/client.go (hunk 3 above)
Comment (Sol-0) [reply pkg/networkservice/common/refresh/client.go:129]: Done

File: pkg/networkservice/common/refresh/client.go (hunk 1 above)
Comment (Sol-0) [reply pkg/networkservice/common/refresh/client.go]: Done

File: pkg/networkservice/common/refresh/client.go (hunk 2 above)
Comment (Sol-0) [reply pkg/networkservice/common/refresh/client.go]: Done

File: pkg/networkservice/common/refresh/client.go (hunk 4 above)
Comment (Sol-0) [reply pkg/networkservice/common/refresh/client.go]: Done

File: pkg/networkservice/common/refresh/client.go (hunk 5 above)
Comment (Sol-0) [reply pkg/networkservice/common/refresh/client.go]: Done

File: pkg/networkservice/common/refresh/client_test.go (hunk 6 above)
Comment (Sol-0) [reply pkg/networkservice/common/refresh/client_test.go]: Done

File: pkg/networkservice/common/refresh/client_test.go (hunk 7 above)
Comment (Sol-0) [reply pkg/networkservice/common/refresh/client_test.go]: Done

**Pull request merged**